From: YOKOTA Hiroshi <yokota.hgml@gmail.com>
Date: Mon, 29 May 2023 23:46:56 +0900
Subject: Revert "PDF Output: Fix regression in previous release that caused
 blank pages when generating headers or footers"

Forwarded: not-needed

This reverts commit 59503c21dc7f59c025d4ace2bfa57643b04dfdab.
---
 src/calibre/utils/podofo/impose.cpp | 13 ++++---------
 1 file changed, 4 insertions(+), 9 deletions(-)

diff --git a/src/calibre/utils/podofo/impose.cpp b/src/calibre/utils/podofo/impose.cpp
index 2923e63..e695fe9 100644
--- a/src/calibre/utils/podofo/impose.cpp
+++ b/src/calibre/utils/podofo/impose.cpp
@@ -6,7 +6,6 @@
  */
 
 #include "global.h"
-#include <sstream>
 #include <string>
 
 using namespace pdf;
@@ -17,14 +16,10 @@ impose_page(PdfMemDocument *doc, unsigned int dest_page_num, unsigned int src_pa
     auto xobj = doc->CreateXObjectForm(src_page.GetMediaBox(), "HeaderFooter");
     xobj->FillFromPage(src_page);
     auto &dest = doc->GetPages().GetPageAt(dest_page_num);
-    dest.GetOrCreateResources().AddResource("XObject", xobj->GetIdentifier(), xobj->GetObject().GetIndirectReference());
-    // prepend the header footer xobject to the stream. This means header/footer is drawn first then the contents, which works
-    // since chromium does not draw in margin areas. The reverse, i.e. appending, does not work with older WebEngine before Qt 6.5.
-    PdfContents *contents = dest.GetContents();
-    std::ostringstream s;
-    s << "q\n1 0 0 1 0 0 cm\n/" << xobj->GetIdentifier().GetString() << " Do\nQ\n" << contents->GetCopy();
-    contents->Reset();
-    contents->GetStreamForAppending().SetData(s.str());
+    PdfPainter painter;
+    painter.SetCanvas(dest);
+    painter.DrawXObject(*xobj, 0, 0);
+    painter.FinishDrawing();
 }
 
 static PyObject*
